File: 2.1.H/Python/habitBased.py
from matchStrategy import MatchStrategy

class HabitBase(MatchStrategy):
    
    # find the individual with cover most habit
    # The habbit is the list of string
    def match(self, individuals, target):

        for individual in individuals:
            if individual.getId() == target.getId():
                individuals.remove(individual)
                break

        individuals.sort(key=lambda individual: self.cover(individual.getHabbit(), target.getHabbit()), reverse=True)
        return individuals

    def cover(self, individual1, individual2):
        cover = 0
        for index, gene in enumerate(individual1):
            if gene in individual2:
                cover += 1
        return cover

File: 2.1.H/Java/StrategyPattern/src/HabitBased.java
import java.util.List;

public class HabitBased implements MatchStrategy{
    @Override
    public List<Individual> match(List<Individual> individuals, Individual targetIndividual) {
        //drop target itself in individuals first
        individuals.remove(targetIndividual);

        // Sort individuals by the number of habits that match with targetIndividual. The habit is List<string>
        individuals.sort((individual1, individual2) -> {
            int count1 = 0;
            int count2 = 0;
            for (String habit : individual1.getHabits()) {
                if (targetIndividual.getHabits().contains(habit)) {
                    count1++;
                }
            }
            for (String habit : individual2.getHabits()) {
                if (targetIndividual.getHabits().contains(habit)) {
                    count2++;
                }
            }
            return count2 - count1;
        });

        individuals.add(0, targetIndividual);
        return individuals;
    }
    
}
